backend/src/org/commcare/session/CommCareSession.java

```diff
@@ -370,6 +394,26 @@ public void setDatum(String keyId, String value) {
         syncState();
     }
 
+    /**
+     * Set a (xml) data instance as the result to a session query datum.
+     * The instance is available in session's evaluation context until the corresponding query frame is removed
+     */
+    public void setQueryDatum(ExternalDataInstance queryResultInstance) {
+        SessionDatum datum = getNeededDatum();
+        if (datum instanceof RemoteQueryDatum) {
+            StackFrameStep step =
+                    new StackFrameStep(SessionFrame.STATE_QUERY_REQUEST,
+                            datum.getDataId(), datum.getValue());
+            frame.pushStep(step);
```

Comment:
The alternative would be to traverse the stack frame in `getEvaluationContext` and pull out any stored instances. Not exactly sure why I took this approach, maybe to avoid needing to add `ExternalDataInstance` externalization logic to `StackFrameStep`.
Do you think storing associated instances in the frame step is better? I guess it does seem conceptually cleaner